File: counting_sort/CountingSort.java
import java.util.Random;

class CountingSort {	//Time complexity = O(n)
    private static void countOccurences(int[] a, int[] c) {	//Basically counting occurences of particular  number and store in its index
        for (int i = 0; i < a.length; i++)	//Counting occurence of each number
            c[a[i]]++;
        for(int i = 1; i < c.length; i++)		//counting total number of numbers less then or equal to that number
            c[i] = c[i-1] + c[i];
    }

    public static int[] countingSort(int[] a, int k) { // For Sorting
        int[] b = new int[a.length];	//b Array stores  sorted array
        int[] c = new int[k+1];		//array  c count occurences
        countOccurences(a, c);
        for (int i = a.length-1; i >= 0; i--) {	//storing number in ascending order in b
            b[c[a[i]]-1] = a[i];	//Stores number to it respective position
            c[a[i]]--;		//Decrease c
        }
        return b;	//Return sorted array
    }

    public static void main(String[] args) {
        int[] A = new int[10000];
        int k = 0;
        Random rand = new Random();
        for (int i = 0; i < A.length; i++) {
            A[i] = rand.nextInt(100);
            if (k < A[i])   //Every number must be between 0 to k
                k = A[i];   //K is max number
        }
        A = countingSort(A, k);
        for (int i = 0; i < A.length; i++)
            System.out.print(A[i] + " ");
    }
}

File: counting_sort/counting_sort.py
#  Running time Complexity: O(z) where z = max(len(array), max(array))
#  Auxillary Space: O(n)


def count_occurences(array, count_array):
    """
    :param array: iterable of elements
    :param count_array: Intermediate variable
    Time Complexity: O(n)
    """
    for i in array:
        count_array[i] += 1
    for i in range(1, len(count_array)):
        count_array[i] += count_array[i - 1]


def counting_sort(array, max_value):
    """
    :param array: Iterable of elements
    :param max_value: Maximum value in array
    :return: Sorted array
    """
    b = [0] * len(array)
    count_array = [0] * (max_value + 1)
    count_occurences(array, count_array)
    for i in range(len(array) - 1, -1, -1):
        b[count_array[array[i]] - 1] = array[i]
        count_array[array[i]] -= 1
    return b


def main():
    array = [3, 4, 5, 3, 6, 3, 3, 5, 5, 3, 2, 1, 5, 8]
    max_value = max(array)
    print(counting_sort(array, max_value))


if __name__ == '__main__':
    main()
